WalletWasabi/Wallets/FilterProcessor/BlockDownloadService.cs

```diff
@@ -0,0 +1,370 @@
+	/// </summary>
+	/// <remarks>
+	/// The method is not efficient but it does not matter too much because the operation is not supposed to be called often.
+	/// The main use case is to deal with a blockchain reorg.
+	/// </remarks>
+	public async Task RemoveBlocksAsync(uint maxBlockHeight)
+	{
+		PriorityQueue<Request, Priority> tempQueue = new(Priority.Comparer);
+
+		List<uint256> toRemoveFromCache = [];
+
+		lock (Lock)
+		{
+			List<(Request Element, Priority Priority)> items = BlocksToDownload.UnorderedItems.ToList();
+
+			foreach ((Request request, Priority priority) in items)
+			{
+				if (priority.BlockHeight <= maxBlockHeight)
+				{
+					tempQueue.Enqueue(request, priority);
+				}
+				else
+				{
+					// The block might have been downloaded by now so just try to set the result.
+					request.Tcs.TrySetResult(new ReorgOccurredResult(NewBlockchainHeight: maxBlockHeight));
+					toRemoveFromCache.Add(request.BlockHash);
+				}
+			}
+
+			BlocksToDownload.Clear();
+			BlocksToDownload.EnqueueRange(tempQueue.UnorderedItems);
+		}
+
+		foreach (uint256 blockHash in toRemoveFromCache)
+		{
+			await FileSystemBlockRepository.RemoveAsync(blockHash, CancellationToken.None).ConfigureAwait(false);
+		}
+	}
+
+	/// <summary>
+	/// Downloads blocks in parallel if there are blocks to download.
+	/// </summary>
+	protected override async Task ExecuteAsync(CancellationToken cancellationToken)
+	{
+		try
+		{
+			List<Task> activeTasks = new(capacity: MaximumParallelTasks);
+
+			while (!cancellationToken.IsCancellationRequested)
+			{
+				bool wait;
+
+				lock (Lock)
+				{
+					// Wait until at least one block-downloading request is here; otherwise, consume requests so that there is at most MAX parallel tasks.
+					wait = BlocksToDownload.Count == 0;
+				}
+
+				if (wait)
+				{
+					await RequestAvailableSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+				}
+
+				lock (Lock)
+				{
+					int toStart = Math.Min(BlocksToDownload.Count, MaximumParallelTasks - activeTasks.Count);
+
+					for (int i = 0; i < toStart; i++)
+					{
+						if (!BlocksToDownload.TryDequeue(out Request? queuedRequest, out _))
+						{
+							throw new UnreachableException("Failed to dequeue block from the queue.");
+						}
+
+						Task newTask = HandleSingleBlockTaskAsync(queuedRequest, cancellationToken);
+						activeTasks.Add(newTask);
+					}
+				}
+
+				// It's still possible that there is no task because a reorg might have occurred.
+				if (activeTasks.Count == 0)
+				{
+					continue;
+				}
+
+				if (activeTasks.Count == MaximumParallelTasks)
+				{
+					Task task = await Task.WhenAny(activeTasks).ConfigureAwait(false);
+					activeTasks.Remove(task);
+				}
+			}
+		}
+		catch (OperationCanceledException)
+		{
+			Logger.LogDebug("Block-downloading service execution was stopped.");
+		}
+		catch (Exception ex)
+		{
+			// This shouldn't happen.
+			Logger.LogError(ex);
+			throw;
```

Comment:
Who will catch this and what will happen then?

Reply:
You are right, we talked about this too many times - I solved it https://github.com/zkSNACKs/WalletWasabi/pull/12328/commits/8b88e6d00debc4f56e6399cd2bbbac3803e66e63

I've just seen your note here.